server/src/main/java/org/opensearch/index/store/RemoteStoreFileDownloader.java

```diff
@@ -0,0 +1,146 @@
+ * pool size and recovery settings.
+ */
+@InternalApi
+public final class RemoteStoreFileDownloader {
+    private final Logger logger;
+    private final ThreadPool threadPool;
+    private final RecoverySettings recoverySettings;
+
+    public RemoteStoreFileDownloader(ShardId shardId, ThreadPool threadPool, RecoverySettings recoverySettings) {
+        this.logger = Loggers.getLogger(RemoteStoreFileDownloader.class, shardId);
+        this.threadPool = threadPool;
+        this.recoverySettings = recoverySettings;
+    }
+
+    /**
+     * Copies the given segments from the remote segment store to the given
+     * local directory.
+     * @param source The remote directory to copy segment files from
+     * @param destination The local directory to copy segment files to
+     * @param toDownloadSegments The list of segment files to download
+     */
+    public void download(Directory source, Directory destination, Collection<String> toDownloadSegments) throws IOException {
+        downloadInternal(source, destination, null, toDownloadSegments, () -> {});
+    }
+
+    /**
+     * Copies the given segments from the remote segment store to the given
+     * local directory, while also copying the segments _to_ another remote directory.
+     * @param source The remote directory to copy segment files from
+     * @param destination The local directory to copy segment files to
+     * @param secondDestination The second remote directory that segment files are
+     *                          copied to after being copied to the local directory
+     * @param toDownloadSegments The list of segment files to download
+     * @param onFileCompletion A generic runnable that is invoked after each file download.
+     *                         Must be thread safe as this may be invoked concurrently from
+     *                         different threads.
+     */
+    public void download(
+        Directory source,
+        Directory destination,
+        Directory secondDestination,
+        Collection<String> toDownloadSegments,
+        Runnable onFileCompletion
+    ) throws IOException {
+        downloadInternal(source, destination, secondDestination, toDownloadSegments, onFileCompletion);
+    }
+
+    private void downloadInternal(
+        Directory source,
+        Directory destination,
+        @Nullable Directory secondDestination,
+        Collection<String> toDownloadSegments,
+        Runnable onFileCompletion
+    ) throws IOException {
+        final Queue<String> queue = new ConcurrentLinkedQueue<>(toDownloadSegments);
+        // Choose the minimum of:
+        // - number of files to download
+        // - max thread pool size
+        // - "indices.recovery.max_concurrent_remote_store_streams" setting
+        final int threads = Math.min(
+            toDownloadSegments.size(),
+            Math.min(threadPool.info(ThreadPool.Names.REMOTE_RECOVERY).getMax(), recoverySettings.getMaxConcurrentRemoteStoreStreams())
+        );
```

Comment:
Curious what is the threadpool queue size to buffer requests if multiple shard recovery is concurrently processing

Reply:
I believe the queue is unbounded, but in practice limited by (number of concurrent recoveries * thread pool size).

Reply:
There might be an issue here - `recoverySettings.getMaxConcurrentRemoteStoreStreams()` would act per shard since the instance is wired up within `IndexShard`.

Reply:
I changed the documentation to reflect the "per recovery" nature of this setting.